ably/realtime_conn_spec_test.go

```diff
@@ -971,6 +972,268 @@ func TestRealtimeConn_RTN15c4(t *testing.T) {
+	// token.
+	var dialURL *url.URL
+	ablytest.Instantly.Recv(t, &dialURL, dials, t.Fatalf)
+
+	if expected, got := "bad:token", dialURL.Query().Get("access_token"); expected != got {
+		t.Errorf("expected reauthorization with token returned by the authCallback; got %q", got)
+	}
+
+	// After a token error response, we finally get to our expected
+	// DISCONNECTED state.
+
+	in <- &proto.ProtocolMessage{
+		Action: proto.ActionError,
+		Error:  &tokenErr,
+	}
+
+	var change ably.ConnectionStateChange
+	ablytest.Instantly.Recv(t, &change, stateChanges, t.Fatalf)
+
+	if change.Current != ably.ConnectionStateDisconnected {
+		t.Fatalf("expected DISCONNECTED event; got %v", change)
+	}
+
+	if change.Reason.StatusCode != tokenErr.StatusCode || change.Reason.Code != tokenErr.Code {
+		t.Fatalf("expected the token error as FAILED state change reason; got: %s", change.Reason)
+	}
+}
+
+func TestRealtimeConn_RTN15h2_Success(t *testing.T) {
+	t.Parallel()
+
+	in := make(chan *proto.ProtocolMessage, 1)
+	out := make(chan *proto.ProtocolMessage, 16)
+
+	dials := make(chan *url.URL, 1)
+
+	c, _ := ably.NewRealtime(ably.ClientOptions{}.
+		Token("fake:token").
+		AutoConnect(false).
+		AuthCallback(func(context.Context, ably.TokenParams) (ably.Tokener, error) {
+			return ably.TokenString("good:token"), nil
+		}).
+		Dial(func(proto string, u *url.URL) (proto.Conn, error) {
+			dials <- u
+			return ablytest.MessagePipe(in, out)(proto, u)
+		}))
+
+	in <- &proto.ProtocolMessage{
+		Action:            proto.ActionConnected,
+		ConnectionID:      "connection-id",
+		ConnectionDetails: &proto.ConnectionDetails{},
+	}
+
+	err := ablytest.ConnWaiter(c, c.Connect, ably.ConnectionEventConnected).Wait()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ablytest.Instantly.Recv(t, nil, dials, t.Fatalf)
+
+	tokenErr := proto.ErrorInfo{
+		StatusCode: 401,
+		Code:       40141,
+		Message:    "fake token error",
+	}
+
+	stateChanges := make(chan ably.ConnectionStateChange, 1)
+
+	off := c.Connection.OnAll(func(change ably.ConnectionStateChange) {
+		stateChanges <- change
+	})
+	defer off()
+
+	in <- &proto.ProtocolMessage{
+		Action: proto.ActionDisconnected,
+		Error:  &tokenErr,
+	}
+
+	// The DISCONNECTED causes a reauth, and dial again with the new
+	// token.
+	var dialURL *url.URL
+	ablytest.Instantly.Recv(t, &dialURL, dials, t.Fatalf)
+
+	if expected, got := "good:token", dialURL.Query().Get("access_token"); expected != got {
+		t.Errorf("expected reauthorization with token returned by the authCallback; got %q", got)
+	}
+
+	// Simulate a successful reconnection.
+	in <- &proto.ProtocolMessage{
+		Action:            proto.ActionConnected,
+		ConnectionID:      "new-connection-id",
+		ConnectionDetails: &proto.ConnectionDetails{},
+	}
+
+	// Expect a UPDATED event.
+
+	var change ably.ConnectionStateChange
+	ablytest.Instantly.Recv(t, &change, stateChanges, t.Fatalf)
+
+	if change.Event != ably.ConnectionEventUpdated {
```

Comment:
The test is failing with a different event here

```
--- FAIL: TestRealtimeConn_RTN15h2_Success (0.00s)
    realtime_conn_spec_test.go:1230: expected UPDATED event; got {DISCONNECTED DISCONNECTED CONNECTED [ErrorInfo :timeout code=80003 statusCode=0] See https://help.ably.io/error/80003}
```

Reply:
This and similar intermittent errors should've been fixed by 324acf8.